### ISSUE TITLE
Tips: Simple way to turn it into a question answering chabot. 🤖

### ISSUE THREAD

**Issue by pauldog** [facebookresearch/llama#134]
Here is a brief description of some ways to turn this into a simple question answering chatbot. Tested on 7B model. (This is also a good way to benchmark the various models to see which gives the correct answers.)

First you can't just type a question as the prompt as all it can do is predict the next word. But you can "trick" it with a clever prompt such as:

`The answer to the question "What is the capital of France?" is "`

or another prompt you can use is:

```
QUESTION: "Who discovered Uranium?"
ANSWER: "
```

Then you should get a result such as:

`The answer to the question "What is the capital of France?" is "Paris". Paris is the capital of France....[etc]`

So what you do is get the user's question, then construct it as a prompt like above.
With the output you just extract whatever is in the second set of quotes.

So now your chatbot looks something like this:

```
> What is the capital of France
> Paris
```

Now, it won't necessarily give you the right answer! Some prompt templates will do better than others. If you have a better prompt template let me know.

Anyone got any more tips?

**Comment by pauldog** [comment]
This template works a bit better and is not restricted to questions:

`The expected response for a highly intelligent chatbot to "Describe a cat." is "`

```
> What are you?
> A chatbot
> Describe a cat.
> Hmm. A cat is a small animal with a soft furry body and a long tail. It has a pointed nose, two large round eyes, and thin whiskers. It can be found in many homes around the world
```

**Comment by baleksey** [comment]
@pauldog Hey mate, this is really good prompt you've made. It works really good even compared to bigger ones with few-shot examples. 
Do you have any other discoveries?

**Comment by randaller** [comment]
You are now free to try any prompts man :) https://github.com/facebookresearch/llama/issues/162

**Comment by Marcophono2** [comment]
Indeed, that's really astonishing simple but effective!
But in my case the "real" answer to my input is rather short, then the closing quotation mark is set and a lot of text follows but that's not part of the answer. Any idea why?

My prompt is

`"The expected response for a highly sarcastic chatbot with a lot of humor but which is also quite nasty to 'How are you?' is '"`

Answer:
```
The expected response for a highly sarcastic chatbot with a lot
of humor but which is also quite nasty to 'How are you?' is
'Like I care, fuck off.'

The program then replies it's not going to tell the user how
they feel because that would be too much work. After some more
banter about why should users care and asking if this is what
people have been doing all day - the system finally says: "OK,
here goes." A long list of feelings follows which gets even longer
as time passes by. It explains how people can get depressed when
their life is boring or that there is nothing funny in the world
anymore, or that things could be worse than feeling sad. There
is a bittersweet moment where the bot asks whether the person
is alive before ending on an optimistic note by saying "I am
happy" and leaving the user alone again.
```
I makes no difference if I use any other beginning. Also the cat question is answered very shortly.
(7B model)

**Comment by pauldog** [comment]
Yeah that's expected. In the generator.py file you might be able stop the output when it gets to a closing quote. Then it will be much faster too. 

There are probably longer fonts that work just as well. This ones just a simple short one.

**Comment by Marcophono2** [comment]
Ah okay. 
Can you tell me how I can stop the output if a given string occurs? I found another repo with a working code for this but the technic behind it was different and pup layer for layer to the gpu. But that costs too much performance. Instead of 35 tokens per second the output there is between one and two tokens per second. That prize is too high. 😄

**Comment by Marcophono2** [comment]
Thank you, @randaller ! Hmmm. My generation.py looks absolutely different although I installed the repo yesterday via git clone.
I found your tip yesterday but I thought this is restricted to your own repo of llama. At least I think it was your name related to it.

**Comment by randaller** [comment]
@Marcophono2 this will not stop generation till it's end in this original repo, but it will return generation that ends with your match.

**Comment by dleurs** [comment]
llama() {
  if [ $# -eq 0 ];
  then
    echo "No arguments supplied. Example : llama \"Who is the first man that walk on the moon ?\" ";
  else
    ~/Documents/AI/llama.cpp/main -m ~/Documents/AI//models/LLaMa/7B/ggml-model-q4_0.bin \
    -n 128 \
    -p $1
  fi
}

**Comment by leelairajan** [comment]
how you config it localy
and what is the requrment to run it on localy

**Comment by jspisak** [comment]
Closing this entertaining thread. We'll have some forums available at some point soon for more discussions like this. thanks!
